### PR TITLE
[GTK] Optimize TreeItem constructor #882

### PR DESCRIPTION
**Pull request by basilevs**
This eliminates a single hotspot. The full complexity for tree building is still quadratic.

Profile before the fix:
Screenshot 2023-11-26 at 21 18 45

Profile after the fix:
Screenshot 2023-11-26 at 21 22 34

### CODE REVIEW

**Comment by SyntevoAlex** [comment]
I would prefer some numbers here in the commit message.
For example:

when inserting 100_000 items into the Tree using this code
```java
Tree tree = new Tree(shell, SWT.VIRTUAL);
for (int i = 0; i < 100_000; i++)
{
    TreeItem item = new TreeItem(tree, 0);
    item.setText("item#" + i);
}
```
the speedup on my machine is 99.99sec -> 88.88sec

File: bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/Tree.java
Comment (SyntevoAlex) [inline bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/Tree.java:203]: Please do not change code style (here, spaces and return on same line).
While I myself find SWT's code style weird, it's still the code style that already exists.

**Review by SyntevoAlex — CHANGES_REQUESTED** [review]
With these small issues fixed, I am willing to merge this PR.

**Comment by SyntevoAlex** [comment]
Please also rename commit to have `[GTK]` prefix, because it only affects GTK. Also, the commit message is poor in the usual way: it answers `what` instead of `why`. A better commit message header would be:
`[GTK] Tree optimization: speedup TreeItem constructor by using known iterator`

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/AllWidgetTests.java
Comment (SyntevoAlex) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/AllWidgetTests.java:36]: It is not appropriate to put performance tests into the main tests package.
Unit tests should run as fast as possible, and performance tests are slow by design.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/performance/Test_org_eclipse_swt_widgets_Tree.java
Comment (SyntevoAlex) [inline tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/performance/Test_org_eclipse_swt_widgets_Tree.java:300]: Asserting on times is unreliable.
If system experience sudden slowdown for a second, tests will fail.
It's fine however if tests are designed to be run manually only.

The other problem is that you're asserting complexity for ALL platforms, not just GTK.
To me it makes sense to convert JUnit test to a manual snippet.

**Review by SyntevoAlex — COMMENTED** [review]
Now that this commit has more issues, I would rather remove it from PR for now.

**Comment by SyntevoAlex** [comment]
With small fixes, commit with the fix is good to merge.
As for the commit with tests, I would rather remove it from PR for now.

I have used up my (rather small) time to look at someone else's PRs this week, so I could only review the next batch in a few days. Feel free to ask someone else to review, but it's not going to be too easy because there're very few active members currently.

If you polish the first commit and don't introduce new problems, I can merge it when I see it.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/performance/Test_org_eclipse_swt_widgets_Tree.java
Comment (basilevs) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/performance/Test_org_eclipse_swt_widgets_Tree.java:300]: I'm trying to fix a performance problem. Note that I do not assert the absolute time and instead check the ratios. There is no way to prevent performance regression besides running the tests in CI. I specifically enabled these tests on CI to be able to have a single point of reference and universal source of timings for the process of this review. The tests are designed to be run on CI.

Checking all the platforms is intentional, as the problem I'm trying to fix affects only Linux and tests should demonstrate it. No platform should exhibit quadratic behaviors, so they also serve as regression tests.

There are already manual snippets in the repository testing performance, but they failed to find the problem in question. I assume nobody runs them.

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/AllWidgetTests.java
Comment (basilevs) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/AllWidgetTests.java:36]: Where should I put these tests to run them on CI?

Comment (SyntevoAlex) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/AllWidgetTests.java:36]: To my knowledge, there is no such place currently.
You can introduce new GitHub job maybe. But frankly I would rather not have it at all. Once problems are fixed, they are unlikely to come back. Yet you are going to burn hours of CPU time over years.

File: bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/Tree.java
Comment (basilevs) [reply bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/Tree.java:203]: Fixed

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/AllWidgetTests.java
Comment (basilevs) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/AllWidgetTests.java:36]: Can leave them in for now and disable before merge to master?

Comment (SyntevoAlex) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/AllWidgetTests.java:36]: You can keep whatever you want on your branch.
You can create another PR for testing.
If you want *this* PR to be merged, remove tests.

Comment (basilevs) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/AllWidgetTests.java]: Removed

File: tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/performance/Test_org_eclipse_swt_widgets_Tree.java
Comment (basilevs) [reply tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/performance/Test_org_eclipse_swt_widgets_Tree.java:300]: Test is now removed from suites and stays passive next to other abandoned performance tests.

**Comment by SyntevoAlex** [comment]
You still have spaces changed in `_getItem`.

**Comment by SyntevoAlex** [comment]
Also, a nitpick: you added unneeded noise in commit message.
Instead of this
```
Before the change:
jfaceReveal[Shape: STAR, virtual: true]
Execution time should grow as 1.100000 degree polynom.
Time for 10000 elements: 620682717.000000 ns
Time for 100000 elements: 62804214489.000000 ns
Ratio: 10.118570
Degree: 2.005119

After the change:
jfaceReveal[Shape: STAR, virtual: true]
Execution time should grow as 1.100000 degree polynom.
Time for 10000 elements: 501378662.000000 ns
Time for 100000 elements: 52781899817.000000 ns
Ratio: 10.527353
Degree: 2.022319
```

I would write
```
In test jfaceReveal[Shape: STAR, virtual: true]:
10_000 elements: 620_682_717ns -> 501_378_662ns
100_000 elements: 62_804_214_489ns -> 52_781_899_817ns
```

Can you see how much easier it is to understand without the noise?
But this is just a rant, I would be pleased to have it changed, but this is not required.

**Comment by basilevs** [comment]
> You still have spaces changed in `_getItem`.
Added spaces before opening parentheses. Hard to spot. Is there a formatter configuration? I could not find one in contribution guide.

**Comment by basilevs** [comment]
> Also, a nitpick: you added unneeded noise in commit message. Instead of this
> I would write

New message:
```
[GTK] Tree optimization: speedup TreeItem constructor by using known
iterator #882

This change eliminates a single hotspot. The overall complexity of
tree building and element reveal is still quadratic.

Performance improvement is proven by running
org.eclipse.swt.tests.junit.performance.Test_org_eclipse_swt_widgets_Tree.jfaceReveal()

In test jfaceReveal[Shape: STAR, virtual: true]:
10_000 elements: 620_682_717ns -> 501_378_662ns
100_000 elements: 62_804_214_489ns -> 52_781_899_817ns (-16%)
```

**Comment by SyntevoAlex** [comment]
Unfortunately you still have spaces changed.
I don't know about formatter configuration. Use some git client that can show edited spaces well. I use SmartGit.

**Review by SyntevoAlex — APPROVED** [review]

**Comment by basilevs** [comment]
> Unfortunately you still have spaces changed.

Fixed two more.

**Comment by SyntevoAlex** [comment]
Oh well. So that you're not upset by spaces, I'll forget it in this commit.
I will however ask you to not change codestyle unnecessarily next time.

**Comment by SyntevoAlex** [comment]
Once CI finishes testing this PR, I will merge it.

**Comment by iloveeclipse** [comment]
Just in case: master is not open yet for merging.

**Comment by SyntevoAlex** [comment]
Ahh, I always forget about that. Thanks for reminding.
Do I understand it correctly that we'll need to wait until 4.30 is released?

**Comment by basilevs** [comment]
@SyntevoAlex thanks for the review.

@iloveeclipse  I've considered merging this in #902 meanwhile, but confused everyone with workflow discussion, so I'm not sure anymore 🤷

**Comment by SyntevoAlex** [comment]
I hope I didn't upset you too much with all the requests. Now that you understand SWT things a bit better, next PR will likely require fewer changes.

**Comment by SyntevoAlex** [comment]
@iloveeclipse while you're around, can I ask you about `assert` usage in SWT?
This PR introduces another `assert` which is the... third! Do you have any ideas why it was avoided before?

**Comment by basilevs** [comment]
> @iloveeclipse while you're around, can I ask you about `assert` usage in SWT? This PR introduces another `assert` which is the... third! Do you have any ideas why it was avoided before?

https://github.com/eclipse-platform/eclipse.platform.swt/blob/518769da51ce22f8242f3feaed522dcf58ebfc4e/bundles/org.eclipse.swt/Readme.md?plain=1#L36-L47

**Comment by SyntevoAlex** [comment]
@basilevs you can now preserve this PR's git branch as is until it's merged.
You can now create another PR (but I won't review it until a few days after because I've hit my limit).
You can rebase the other PR on top of this branch (so that it will contain copies of commits from this PR).
When this PR is merged, the other PR will become based on master.

**Comment by iloveeclipse** [comment]
> @iloveeclipse while you're around, can I ask you about `assert` usage in SWT? This PR introduces another `assert` which is the... third! Do you have any ideas why it was avoided before?

I can't say why it wasn't used but I personally prefer not to use it too, so I would prefer we would not introduce it. I've not seen yet code that would really need that ancient C feature that was by mistake added to Java.

Thinking more about it, I believe SWT was created before this mistake happened to Java.

**Comment by SyntevoAlex** [comment]
Why do you consider it a mistake?
I myself like to add asserts whenever appropriate, and they actually catch quite a few bugs,

**Comment by iloveeclipse** [comment]
In most places I saw an assert it was a lazy way to say "actually we should check that and somehow handle possible wrong flow, but let it go wrong in production code, it can't really happen". So either it can't happen and assert is not needed, or the code will run with wrong state in production with side effects we haven't tested, because in tests we use assertions and flow is different anyway. So either one uses defensive code style, and no assertions needed because all possible code paths are handled by regular code flow, or one uses assertions.

**Comment by SyntevoAlex** [comment]
Yes I can understand this logic.

However, for every possible assert location, let's estimate the likelihood of things going wrong.
If likelihood is high, then proper error handling code is needed. But handling errors properly is difficult, so there's a cost with it.
If likelihood is low, yet still your gut feeling is worried, that's a good place for assert. It's better than nothing, and cheap enough.
If likelihood is next to zero, then it's fine to not add any handling.

So my point is that there are many cases where paying the cost for proper handling is not justified because the problem seems to be very unlikely, yet still I'm not 100% sure.

**Comment by laeubi** [comment]
If I assume/assert something can't happen I throw `AssertionError` with a message why I think it can't happen, that is independent of if assertions are enabled or not if throwing an exception is not appropriate. Right after that I most of the time delete it and handle it in a proper way... because according to [Murphy's law](https://en.wikipedia.org/wiki/Murphy%27s_law) "Anything that can go wrong will go wrong, and at the worst possible time." :-)

**Comment by SyntevoAlex** [comment]
master is not open. Please rebase your PR onto master so that tests are run, then I can merge it.

**Comment by SyntevoAlex** [comment]
In the future, don't merge master into your branch. Instead, rebase your branch onto master.

**Comment by SyntevoAlex** [comment]
In master now. Thanks for your contribution!

You can now rebase your other branch onto master.

**Comment by basilevs** [comment]
Next optimization step is https://github.com/eclipse-platform/eclipse.platform.swt/pull/904

**Pull request merged**
